Don't set MSBuildAllProjects for MSBuild 16.0+

drewnoakes opened the pull request:
Microsoft/msbuild#1299 discusses a change whereby MSBuild prepends the newest project file path to `MSBuildAllProjects` automatically. Therefore, so long as consumers of this property only use this property for up-to-date checks, it's possible to reduce both the size of this property and corresponding allocations.

This reduction applies equally to consumers, such as for the case discussed in dotnet/project-system#3744.

For backwards compatibility the property is still set for MSBuild versions prior to 16.0.

See also Microsoft/msbuild#3605.

//cc: @jeffkl @rainersigwald @davkean

drewnoakes committed:
Don't set MSBuildAllProjects

Microsoft/msbuild#1299 discusses a change whereby MSBuild prepends the newest input file path to this property automatically. Therefore, so long as consumers of this property only use this list for up-to-date checks, it's possible to reduce the size of this property, and reduce corresponding allocations.

This reduction applies equally to consumers, such as in the case discussed in dotnet/project-system#3744

For backwards compatibility the property is still set for MSBuild versions prior to 16.0.

livarcocc commented on src/Assets/TestProjects/AppWithSharedProject/SharedProject/SharedProject.projitems:
```diff
@@ -1,7 +1,9 @@
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+    <MSBuildAllProjects Condition=" '$(MSBuildVersion)' == '' Or '$(MSBuildVersion)' &lt; '16.0' ">
```
I am not sure if this has been discussed. But these files should not be used with MSBuild less than 16.0.

The CLI will require Dev16, which means if someone tries to use this in Dev15, they will have to do hacks to enable it and we don't support it.

And the CLI itself will carry MSBuild 16.

cc @rainersigwald

drewnoakes commented:
> here was my commit that I hadn't had a change to submit yet

If it's safe to remove them outright then that's a better approach, and given your commit we can close this PR.

> I'm surprised you didn't have to update the unit test like I had to.

It's marked as `Skip`. It does look like it should be removed, as you did.

jeffkl commented:
My commit is from August and I haven't had time to send a PR so feel free to merge this PR instead.

drewnoakes commented:
In that case I'll remove the skipped unit test and remove the `< 16.0` checks altogether.

drewnoakes committed:
Remove adding to MSBuildAllProjects altogether

Usages of these files will at least run MSBuild 16.0 so the backwards
compatibility is not required.

drewnoakes committed:
Remove unit test for MSBuildAllProjects

rainersigwald reviewed (commented):
🎉

I closed https://github.com/dotnet/sdk/issues/1810 since this will make it super obsolete.

davkean commented:
> The CLI will require Dev16, which means if someone tries to use this in Dev15, they will have to do hacks to enable it and we don't support it.

How is that enforced? ie What will prevent this from "just working" if I run MSBuild.exe (v15) against a project that only has 3.0 installed.

livarcocc commented:
@davkean the 3.0 CLI has a file that establishes 16.0 as the minimum MSBuild version it requires. This file is then honored by the SDK Resolver that msbuild uses to find the .NET Core SDK props/targets. Users would have to edit this file inside the SDK installation to get around that. At that point, I would say all bets are OFF.

davkean commented:
👍
